V1 Implementation of Adaptive Environments (Tunnels) + Bezier Curves

SumeetBatra opened the pull request:
Infrastructure code that allows us to change the shape and size of the environment during training. Can use this code to build rules in the future that decide how the environment changes over time based on statistics of the last N episodes, etc. 

EDIT: Added Bezier Curves for evader-pursuit task

SumeetBatra committed:
minor bug fixes

SumeetBatra committed:
merge with upstream

SumeetBatra committed:
Merge branch 'multi-quad' of https://github.com/amolchanov86/gym_art into multi-quad

SumeetBatra committed:
support for adaptive environments. Can now specify room length, width, and height. Changes to these dims will be reflected in quadrotor dyanmics and in the visualization at test time

SumeetBatra committed:
simplify random state calc

SumeetBatra committed:
code cleanup

SumeetBatra committed:
added bezier curves for evader-pursuit task

SumeetBatra committed:
increased default size of env (apparently this is harder w/ bezier curve-following goals)

alex-petrenko commented on gym_art/quadrotor_multi/quadrotor_multi.py:
```diff
@@ -23,12 +24,12 @@ def __init__(self,
                  dynamics_randomize_every=None, dyn_sampler_1=None, dyn_sampler_2=None,
                  raw_control=True, raw_control_zero_middle=True, dim_mode='3D', tf_control=False, sim_freq=200.,
-                 sim_steps=2, obs_repr='xyz_vxyz_R_omega', ep_time=7, obstacles_num=0, room_size=10,
+                 sim_steps=2, obs_repr='xyz_vxyz_R_omega', ep_time=7, obstacles_num=0, room_length=20, room_width=20, room_height=20,
```
Any reason to increase the size of the room initially?
I think 10 is a good value.

Or was 10 like a "radius" of the room, rather than the length of 1 side?

SumeetBatra replied:
10 is the length of one side. I changed it to 20 just to give enough space for the drones to follow the bezier curves. It looked like the drones were going pretty close to the edge at just 10.

alex-petrenko commented on gym_art/quadrotor_multi/quadrotor_multi_visualization.py:
```diff
@@ -13,8 +13,8 @@
-            quad_arm=None, models=None, obstacles=None, visible=True, resizable=True, goal_diameter=None,
-            viewpoint='chase', obs_hw=None, obstacle_mode='no_obstacles'
+            quad_arm=None, models=None, obstacles=None, visible=False, resizable=True, goal_diameter=None,
+            viewpoint='chase', obs_hw=None, obstacle_mode='no_obstacles', room_dims=(10, 10, 10)
```
Here it's 10, 10, 10, but in the ctor it's 20, 20, 20
Let's keep it at 10,10,10 maybe?

SumeetBatra replied:
Maybe we should keep default to 20x3 for everything? just because while we are developing new methods and running experiments it might be better not to have to deal with the room bounds.

alex-petrenko replied:
Don't care much either way, but I honestly don't see a reason to change this. The current value seems to work, why change it then? The first rule of programming

Another concern is that bigger numerical values like 20 and -20 might be hard for neural networks to work with, although with tanh activations that we use this might not be a problem

alex-petrenko commented on gym_art/quadrotor_multi/rendering3d.py:
```diff
@@ -604,6 +604,14 @@ def box(x, y, z):
+# Box representing the environment. Shortcut way to add walls during visualization
+def envBox(x, y, z):
+    corner1 = np.array([-x, -y, z])
+    corner2 = np.array([-x, y, 0])
```
Should this be [x, y, 0]?

alex-petrenko commented on gym_art/quadrotor_multi/rendering3d.py:
```diff
@@ -670,6 +679,20 @@ def box_mesh(x, y, z):
+    vbottom = deepcopy(vtop)
+    vbottom[:, 2] = 0
+    v = np.concatenate([vtop, vbottom], axis=0)
+    # WTF IS THIS?????? lol
```
You have 8 vertices (4 points in vtop and 4 points in vbottom).
t is an array of triangles, for some reason they are 1-indexed (and then 1 is subtracted in the end, but ok, whatever)

To invert normals you will need to change the order of the vertices in triangles, as we discussed.
So 1,3,2 should become 1,2,3, etc... (I believe)

SumeetBatra committed:
fixed issue with bezier curves where goal would sample too far away in large rooms resulting in absurd goal velocities that the drones can't handle

SumeetBatra committed:
GUI fix for Retina Displays + Room Visualizations
fixed issue where only one-fourth the window would display when visualizing the simulator on retina displays.
Inverted normals on the triangles for visualizing the room walls.

SumeetBatra committed:
fix spaces

SumeetBatra committed:
fixed issue where goal would keep sampling only upwards w/ new logic

Commit:
Merge branch 'multi-quad' into multi-quad
